mslib/mswms/mss_plot_driver.py

```diff
@@ -676,3 +677,241 @@ def plot(self):
+        # file is open).
+        lon_data = ((self.lon_data - left_longitude) % 360) + left_longitude
+        lon_indices = lon_data.argsort()
+        lon_data = lon_data[lon_indices]
+        factors = []
+
+        for name, var in self.data_vars.items():
+            data[name] = []
+            if len(var.shape) == 4:
+                var_data = var[timestep, ::-self.vert_order, ::self.lat_order, :]
+            else:
+                var_data = var[:][timestep, np.newaxis, ::self.lat_order, :]
+            logging.debug("\tLoaded %.2f Mbytes from data field <%s> at timestep %s.",
+                          var_data.nbytes / 1048576., name, timestep)
+            logging.debug("\tVertical dimension direction is %s.",
+                          "up" if self.vert_order == 1 else "down")
+            logging.debug("\tInterpolating to cross-section path.")
+            # Re-arange longitude dimension in the data field.
+            var_data = var_data[:, :, lon_indices]
+
+            points = (self.vert_data, self.lat_data, lon_data)
+            inter_points = np.array([[self.alts[i], self.lats[i], self.lons[i]] for i in range(len(self.lats))])
+            interpolator = RegularGridInterpolator(points, var_data, bounds_error=False)
+            if name == "air_pressure":
+                for index, point in enumerate(inter_points):
+                    vertical_slice = np.array(
+                        [[self.vert_data[i], point[1], point[2]] for i in range(len(self.vert_data))])
+                    pressures = interpolator(vertical_slice)
+                    closest = 0
+                    direction = 1
+                    for i, pressure in enumerate(pressures):
+                        if abs(pressure - point[0]) < abs(pressures[closest] - point[0]):
+                            closest = i
+                            direction = 1 if pressure - point[0] > 0 else -1
+                    next_closest = closest + direction if closest < len(pressures) - 1 else closest
+                    distance = abs(pressures[closest] - point[0]) + abs(pressures[next_closest] - point[0])
+                    factors.append(
+                        [[closest, abs(pressures[closest] - point[0]) / distance],
+                         [next_closest, abs(pressures[next_closest] - point[0]) / distance]])
+
+            for index, point in enumerate(inter_points):
+                cur_factor = factors[index]
+                vertical_slice = np.array(
+                    [[self.vert_data[i], point[1], point[2]] for i in range(len(self.vert_data))])
+                vertical = interpolator(vertical_slice)
+                value = vertical[cur_factor[0][0]] * cur_factor[0][1] + \
+                    vertical[cur_factor[1][0]] * cur_factor[1][1]
+                data[name].append(value)
+
+            # Free memory.
+            del var_data
+            data[name] = np.array(data[name])
+
+        return data
+
+    def shift_data(self):
```

Comment:
can we factor this out somehow? This code is fickle (fixed a bug only recently) and having it in multiple places is not nice.